include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +174,738 @@ struct Basic
+    scalar_type getLdotH() NBL_CONST_MEMBER_FUNC { return iso_cache.LdotH; }
+    scalar_type getNdotH() NBL_CONST_MEMBER_FUNC { return iso_cache.NdotH; }
+    scalar_type getNdotH2() NBL_CONST_MEMBER_FUNC { return iso_cache.NdotH2; }
+
+    scalar_type getTdotH() NBL_CONST_MEMBER_FUNC { return TdotH; }
+    scalar_type getBdotH() NBL_CONST_MEMBER_FUNC { return BdotH; }
+
+    isocache_type iso_cache;
+    scalar_type TdotH;
+    scalar_type BdotH;
 };
 
 
-// finally fixed the semantic F-up, value/pdf = quotient not remainder
-template<typename SpectralBins>
-struct quotient_and_pdf
+#define NBL_CONCEPT_NAME BxDF
+#define NBL_CONCEPT_TPLT_PRM_KINDS (typename)
+#define NBL_CONCEPT_TPLT_PRM_NAMES (T)
+#define NBL_CONCEPT_PARAM_0 (bxdf, T)
+#define NBL_CONCEPT_PARAM_1 (spec, typename T::spectral_type)
+#define NBL_CONCEPT_PARAM_2 (pdf, typename T::scalar_type)
+#define NBL_CONCEPT_PARAM_3 (_sample, typename T::sample_type)
+#define NBL_CONCEPT_PARAM_4 (iso, typename T::isotropic_interaction_type)
+#define NBL_CONCEPT_PARAM_5 (aniso, typename T::anisotropic_interaction_type)
+#define NBL_CONCEPT_PARAM_6 (param_iso, typename T::params_isotropic_t)
+#define NBL_CONCEPT_PARAM_7 (param_aniso, typename T::params_anisotropic_t)
+#define NBL_CONCEPT_PARAM_8 (u, vector<typename T::scalar_type, 2>)
+NBL_CONCEPT_BEGIN(9)
+#define bxdf NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_0
+#define spec NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_1
+#define pdf NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_2
+#define _sample NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_3
+#define iso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_4
+#define aniso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_5
+#define param_iso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_6
+#define param_aniso NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_7
+#define u NBL_CONCEPT_PARAM_T NBL_CONCEPT_PARAM_8
+NBL_CONCEPT_END(
+    ((NBL_CONCEPT_REQ_TYPE)(T::scalar_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::isotropic_interaction_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::anisotropic_interaction_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::sample_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::spectral_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::quotient_pdf_type))
+    ((NBL_CONCEPT_REQ_TYPE)(T::params_isotropic_t))
+    ((NBL_CONCEPT_REQ_TYPE)(T::params_anisotropic_t))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.eval(param_iso)), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.eval(param_aniso)), ::nbl::hlsl::is_same_v, typename T::scalar_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.generate(iso,u)), ::nbl::hlsl::is_same_v, typename T::sample_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.generate(aniso,u)), ::nbl::hlsl::is_same_v, typename T::sample_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.quotient_and_pdf(param_iso)), ::nbl::hlsl::is_same_v, typename T::quotient_pdf_type))
+    ((NBL_CONCEPT_REQ_EXPR_RET_TYPE)((bxdf.quotient_and_pdf(param_aniso)), ::nbl::hlsl::is_same_v, typename T::quotient_pdf_type))
```

Comment:
why do `eval` and `quotient_and_pdf` have some weird parameter struct instead of taking interaction, sample and cache?